File: shipbob_to_gorgias/java/src/test/java/sb2gorgias/CronTest.java
package sb2gorgias;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;

class CronTest {

    /** A ShipBob-shaped timestamp days back - seven fractional digits, as the real API
     * sends - inside clamp's 30-day window. */
    private static String ago(int days) {
        OffsetDateTime dt = OffsetDateTime.now(ZoneOffset.UTC).minusDays(days);
        return dt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss")) + ".1234567+00:00";
    }

    @Test
    void clampBoundsCursorBetweenOneMonthAgoAndNow() {
        OffsetDateTime now = OffsetDateTime.of(2026, 7, 16, 12, 0, 0, 0, ZoneOffset.UTC);
        assertEquals(OffsetDateTime.of(2026, 7, 10, 0, 0, 0, 0, ZoneOffset.UTC),
                Cron.clamp("2026-07-10T00:00:00", now)); // in range
        assertEquals(now, Cron.clamp("2099-01-01T00:00:00", now)); // future -> now
        assertEquals(now.minus(Cron.ONE_MONTH), Cron.clamp(null, now)); // missing -> floor
    }

    @Test
    void runPagesUntilEmptyUpsertsCustomerAndAdvancesCursor() {
        FakeShipBobClient shipbob = new FakeShipBobClient(
                List.of(List.of(
                        Helpers.makeOrder(1, ago(6), "jane@example.com"),
                        Helpers.makeOrder(2, ago(5), "jane@example.com")
                )),
                List.of(),
                null
        );
        FakeGorgiasClient gorgias = new FakeGorgiasClient();
        Pandium pandium = Helpers.pandium().secrets(Helpers.GORGIAS_SECRETS)
                .config(Map.of("order_start_date", ago(20))).build();

        JSONObject record = Cron.runCron(pandium, new Cron.Deps(shipbob, gorgias, new FakeWatchdogArmer(),
                code -> { throw new SimulatedExit(code); }, OffsetDateTime.now(ZoneOffset.UTC)));

        assertEquals(List.of(1, 2), shipbob.pages.get("new")); // paged until the empty page
        assertEquals(1, gorgias.createLog.size()); // both orders batch onto one customer
        assertEquals(Util.trimTo(ago(5), 26), record.getString("new_order_start_date")); // advanced to the last order

        JSONObject finalCustomer = gorgias.updateLog.get(gorgias.updateLog.size() - 1).getValue();
        var finalOrders = finalCustomer.getJSONObject("data").getJSONObject("pandium").getJSONArray("shipbob_orders");
        List<Long> ids = new java.util.ArrayList<>();
        for (int i = 0; i < finalOrders.length(); i++) {
            ids.add(finalOrders.getJSONObject(i).getLong("id"));
        }
        ids.sort(null);
        assertEquals(List.of(1L, 2L), ids);
    }

    /** Pages are each sorted newest-first, but not relative to each other, so the cursor has
     * to be the oldest update seen anywhere - not the last one processed. */
    @Test
    void runAdvancesUpdatedCursorToOldestUpdateAcrossPages() {
        FakeShipBobClient shipbob = new FakeShipBobClient(
                List.of(),
                List.of(
                        List.of(Helpers.makeOrder(1, ago(2), "j@x.com"), Helpers.makeOrder(2, ago(3), "j@x.com")),
                        List.of(Helpers.makeOrder(3, ago(9), "j@x.com"), Helpers.makeOrder(4, ago(8), "j@x.com")),
                        List.of(Helpers.makeOrder(5, ago(4), "j@x.com"))
                ),
                null
        );
        Pandium pandium = Helpers.pandium().secrets(Helpers.GORGIAS_SECRETS)
                .config(Map.of("order_start_date", ago(20))).build();

        JSONObject record = Cron.runCron(pandium, new Cron.Deps(shipbob, new FakeGorgiasClient(), new FakeWatchdogArmer(),
                code -> { throw new SimulatedExit(code); }, OffsetDateTime.now(ZoneOffset.UTC)));

        String expected = Cron.formatCursor(Util.parseTimestamp(ago(9)).orElseThrow());
        assertEquals(expected, record.getString("updated_order_start_date")); // not order 5, the last processed
    }

    /** The two cursors resume differently. new_order_start_date climbs per order over an
     * oldest-first query, so it is sound wherever the run stops. updated_order_start_date is
     * the minimum across every page, so it only holds once the query is exhausted - an unread
     * page can carry an older update - and a run cut short flushes the value it started
     * with. */
    @Test
    void timeoutFlushesTheFinishedHalfAndLeavesTheInterruptedOne() {
        String start = ago(20);
        FakeWatchdogArmer armer = new FakeWatchdogArmer();
        FakeShipBobClient shipbob = new FakeShipBobClient(
                List.of(List.of(Helpers.makeOrder(1, ago(6), "j@x.com"))),
                List.of(
                        List.of(Helpers.makeOrder(2, ago(2), "j@x.com")),
                        List.of(Helpers.makeOrder(3, ago(9), "j@x.com")) // never read
                ),
                (half, page) -> {
                    if ("updated".equals(half) && page == 2) {
                        armer.fire();
                    }
                }
        );
        Pandium pandium = Helpers.pandium().secrets(Helpers.GORGIAS_SECRETS)
                .config(Map.of("order_start_date", start)).build();

        ByteArrayOutputStream capturedOut = new ByteArrayOutputStream();
        PrintStream originalOut = System.out;
        System.setOut(new PrintStream(capturedOut, true, StandardCharsets.UTF_8));
        SimulatedExit thrown = null;
        try {
            Cron.runCron(pandium, new Cron.Deps(shipbob, new FakeGorgiasClient(), armer,
                    code -> { throw new SimulatedExit(code); }, OffsetDateTime.now(ZoneOffset.UTC)));
        } catch (SimulatedExit e) {
            thrown = e;
        } finally {
            System.setOut(originalOut);
        }

        assertNotNull(thrown); // a timed-out run still succeeds, so progress merges
        assertEquals(0, thrown.code);
        String[] lines = capturedOut.toString(StandardCharsets.UTF_8).strip().split("\n");
        JSONObject flushed = new JSONObject(lines[lines.length - 1]);
        assertEquals(Util.trimTo(ago(6), 26), flushed.getString("new_order_start_date")); // that half finished
        assertEquals(Util.trimTo(start, 26), flushed.getString("updated_order_start_date")); // this one did not
    }
}

File: shipbob_to_gorgias/kotlin/app/src/test/kotlin/sb2gorgias/CronTest.kt
package sb2gorgias

import java.time.LocalDateTime
import java.util.concurrent.atomic.AtomicReference
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlinx.serialization.json.JsonPrimitive

/** Flow A: the resumable order sync. */
class CronTest {
    private fun at(value: String): LocalDateTime = requireNotNull(parseTimestamp(value)) { "a test timestamp" }

    private fun cursorsFrom(start: String) = AtomicReference(Cursors(at(start), at(start)))

    /**
     * Run the sync the way a tenant with the default settings would: oldest first, and a
     * `now` comfortably later than every timestamp in the fixtures.
     */
    private fun runSync(shipbob: Orders, gorgias: Helpdesk, cursors: AtomicReference<Cursors>) =
        sync(shipbob, gorgias, cursors, newestFirst = false, now = at("2026-07-20T00:00:00"))

    private fun updatedOn(id: Long, day: Int) =
        orderUpdatedOn(id, "2026-07-%02dT00:00:00Z".format(day), "j@x.com")

    @Test
    fun `the sync pages until empty and keeps the cursor current as it goes`() {
        // Advancing the cursor per order rather than once at the end is what makes the
        // flow resumable.
        val shipbob =
            FakeShipBob(
                newPages =
                    listOf(
                        listOf(order(1, "2026-07-05T10:00:00Z", "j@x.com")),
                        listOf(order(2, "2026-07-06T10:00:00Z", "j@x.com")),
                    ),
            )
        val gorgias = RecordingGorgias()
        val cursors = cursorsFrom("2026-07-01")
        shipbob.watched = cursors

        runSync(shipbob, gorgias, cursors)

        assertEquals(listOf(1, 2, 3), shipbob.newPagesRequested) // until empty

        // What a flush would have written when page 2 was fetched: order 1 done.
        assertEquals("2026-07-05T10:00:00", shipbob.newCursorWhenFetched[2]?.let(::isoTimestamp))
        assertEquals(
            JsonPrimitive("2026-07-06T10:00:00"),
            cursors.get().asMetadata()["new_order_start_date"],
        )

        // Both orders batch onto one customer: created once, then updated.
        assertEquals(1, gorgias.created.size)
        val (_, customer) = gorgias.updated.last()
        assertEquals(2, customer["data"]["pandium"]["shipbob_orders"].list.size)
    }

    @Test
    fun `the updated cursor lands on the oldest update across every page`() {
        // Pages are each sorted newest-first, but not relative to each other, so the
        // cursor has to be the oldest update seen anywhere.
        val shipbob =
            FakeShipBob(
                updatedPages =
                    listOf(
                        listOf(updatedOn(1, 18), updatedOn(2, 17)),
                        listOf(updatedOn(3, 11), updatedOn(4, 12)), // the oldest update overall
                        listOf(updatedOn(5, 16)), // newer again, after the oldest page
                    ),
            )
        val cursors = cursorsFrom("2026-07-01")
        shipbob.watched = cursors

        runSync(shipbob, RecordingGorgias(), cursors)

        assertEquals(
            JsonPrimitive("2026-07-11T00:00:00"), // not order 5, the last one processed
            cursors.get().asMetadata()["updated_order_start_date"],
        )

        // Until the last page is in the minimum is provisional, so a flush partway
        // through writes the value the run started with.
        for (page in 1..3) {
            assertEquals(
                "2026-07-01T00:00:00",
                shipbob.updatedCursorWhenFetched[page]?.let(::isoTimestamp),
                "the cursor moved while page $page was still outstanding",
            )
        }
    }

    @Test
    fun `a page that fails to fetch ends the run rather than committing a cursor`() {
        // A page that errors has to stay distinguishable from the empty page that ends
        // the loop, or the run would stop early and commit a cursor for pages it never
        // read.
        val shipbob =
            FakeShipBob(
                updatedPages =
                    listOf(
                        listOf(updatedOn(1, 18)),
                        listOf(updatedOn(2, 11)), // never read: the fetch fails first
                    ),
                failingPage = 2,
            )
        val cursors = cursorsFrom("2026-07-01")

        assertFailsWith<IllegalStateException>("a failed fetch is not an exhausted query") {
            runSync(shipbob, RecordingGorgias(), cursors)
        }

        // Stopped at the failure rather than paging on.
        assertEquals(listOf(1, 2), shipbob.updatedPagesRequested)
    }
}
